frontend/catalyst/api_extensions/callbacks.py

```diff
@@ -39,6 +40,95 @@
 from catalyst.utils.types import convert_pytype_to_shaped_array
 
 
+class CallbackWithCustomGrad:
+    """A callback with a custom grad"""
+
+    def __init__(self, func, restype, forward, reverse):
+        assert func and forward and reverse
+        self.func = func
+        self.restype = restype
+        self._fwd = forward
+        self._fwd_jaxpr = None
+        self._bwd = reverse
+        self._bwd_jaxpr = None
+        self.callback = None
+
+    def __call__(self, *args, **kwargs):
+        if self.callback:
+            return self.callback(*args, **kwargs)
+
+        def closure(*args, **kwargs) -> self.restype:
+            return self.func(*args, **kwargs)
```

Comment:
No, it is not necessarily annotated. We use an alternative functional syntax in that case.

Reply:
Can we replace `func` at the point where we detect the functional syntax then? That way we don't need to replicate this wrapper in a few places and the logic for it is more localized. 
I'm also finding the name somewhat confusing because we don't appear to close over any variables besides the type annotation.

Reply:
Can we get this in and then I can clean this bit up? I just don't want to add any extra time. This seems like a really small issue that (while I agree can be solved here) I just want to get this feature in. I also need time to work on QSVT.

Reply:
Tbh the smaller the issue the less likely it is going to be picked in the future. I already forgot the issues that were left for later even in just this series of PRs and I doubt we'll ever come back to those either.

But I won't reject the PR over it.